tools/link.py

```diff
@@ -220,35 +220,37 @@ def get_js_sym_info():
     input_files.append(read_file(jslib))
   content = '\n'.join(input_files)
   content_hash = hashlib.sha1(content.encode('utf-8')).hexdigest()
+  library_syms = None
 
-  def build_symbol_list(filename):
-    """Only called when there is no existing symbol list for a given content hash.
-    """
-    library_syms = generate_js_sym_info()
-
-    write_file(filename, json.dumps(library_syms, separators=(',', ':'), indent=2))
-
-  # We need to use a separate lock here for symbol lists because, unlike with system libraries,
-  # it's normally for these file to get pruned as part of normal operation.  This means that it
-  # can be deleted between the `cache.get()` then the `read_file`.
-  with filelock.FileLock(cache.get_path('symbol_lists.lock')):
-    filename = cache.get(f'symbol_lists/{content_hash}.json', build_symbol_list)
-    library_syms = json.loads(read_file(filename))
-
-    # Limit of the overall size of the cache.
-    # This code will get test coverage since a full test run of `other` or `core`
-    # generates ~1000 unique symbol lists.
-    cache_limit = 500
-    root = cache.get_path('symbol_lists')
-    if len(os.listdir(root)) > cache_limit:
+  cache_file = str(cache.get_path(f'symbol_lists/{content_hash}.json'))
+
+  utils.safe_ensure_dirs(cache.get_path('symbol_lists'))
+  with filelock.FileLock(cache_file + '.lock'):
+    if os.path.exists(cache_file):
+      # Cache hit, read the file
+      library_syms = json.loads(read_file(cache_file))
+    else:
+      # Cache miss, generate the symbol list and write the file
+      library_syms = generate_js_sym_info()
```

Comment:
The description says

> Prior to this change the symbol generation process itself always happened with the symbol list lock held. Now in the most common cases the symbol list is generated before we grab the lock.

But this line is run under the lock from 228?

It looks like only the cache limit check was moved out of the lock - was that really so slow?

Reply:
But now we use a per-file lock, not a global lock, so only processes with the exact same linker flags will be in contention here.

Reply:
I update the description.